Domains: ccTLDs/ca-form: Always show organization field

deBhal opened the pull request:
This PR fixes a hole in #20984 where organization validation errors can be present while the organization field is hidden, leading to misleading error notifications and failed domain purchases.

Originally these improvements where in that PR, but I'm splitting them out here in order to deploy the approved improvements sooner.

To test:
1) Add a .ca domain to your cart
2) fill out the main form and continue to the .ca details form
3) Make sure that validation errors on the organization field are handled correctly
 - Can't submit invalid organizations
 - Correct validation errors are shown for both .CA specific validation (whitelist, required for CCO) and general org name validation (limited special characters, length)

I also added an `optional` option to the form-label component that mirrors the `required` option. I refactored the fr-form to use it, so that's worth checking that. It's not used by anything else, but it wouldn't hurt to have a bit of a general check of form labels.

Original comment from #20984:
Ah, I see. Good catch!

Ironically, the "required" check is the one that I do handle, and it's the other checks  that trigger that error, despite what that error notification claims (invalid characters, too long or the .ca specific naming).

Those are real validation errors that we don't want to ignore. OpenSRS doesn't care, but we don't want them in the database.

I first tried showing the field if there's an error for it and that works fine, but it feels bad. You fix the problem, and the field inexplicably disappears, even if you haven't finished changing the value there. Clearing the field would also work, but that's a weird exception to add.

So I think the only way that hangs together is to always have the field visible, and use the secondary cues to guide the user - enabling/disabling the checkout button, the error messages, and the "optional" decoration:

![ca-form- always show organization field](https://user-images.githubusercontent.com/5952255/36770501-43843782-1c96-11e8-938f-94bf5ab34f1d.gif)
#
One minor imperfection is that we do run the org name validation for non-corporation entities, even though OpenSRS doesn't care. This is Consistent with the main domain details form, and a pretty minor inconvenience.

One new thing to check: While I was de-duping the "optional" styling I found the `required` prop in the FormLabel component, so I've added optional there alongside it, as it seems like something with pretty broad applicability. I `grep`ed the codebase to make sure that `optional` is not being used as a prop anywhere, so the only thing that this will effect is the fr-form, which should look exactly the same. For your convenience, here's a before pic to compare:

![fr-form-before](https://user-images.githubusercontent.com/5952255/34429935-a1d5338e-ecaa-11e7-8014-4e9e74f0a109.jpg)

Another thing worth noting is that the organization field in both the `domain-details-form` and the `ca-form` are validating against all tlds. If you add a `(.*)?.uk` domain to the cart, you can get the .uk specific error about corporation names being at least 4 characters:

![banners_and_alerts_and_checkout_ _julesauspremiumtest_ _wordpress_com](https://user-images.githubusercontent.com/5952255/36770641-27326774-1c97-11e8-9192-fd66f098cb1a.jpg)
( "Numele organizației trebuie să aibă cel puțin 4 caractere." )

ramonjd commented on client/components/domains/registrant-extra-info/ca-form.jsx:
```diff
@@ -160,24 +165,19 @@ class RegistrantExtraInfoCaForm extends React.PureComponent {
-
-		const formIsValid = ciraAgreementAccepted && organizationFieldIsValid;
+		const organizationFieldProps = this.addRequiredOrganizationValidation(
+			this.props.getFieldProps( 'organization', true )
```
Aside from the imminent changes introduced by #19889, I'm not sure it's a good idea to pass state controllers between forms. The most obvious case I can think of is when someone decides to change the properties of the originating field in `domain-details-form` for whatever reason. I know it's unlikely for now, but I'm just being overly cautious.

Also `ca-form`, just like `fr-form`, is a self-contained form under `<ExtraInfoForm />` with idiosyncratic validation anyway. If we have to embed further regulatory checks later, managing the form's state internally might make it easier to test in isolation.

If we need the same validation across forms, we could abstract the validator function itself, or pass it as a prop. And we can already grab/update contact details in the store. 

What do you think?

deBhal commented:
#22156 removes the need for the extra props here, so after that lands I should be able to tidy this up nicely and wrap it up.

ramonjd commented:
> #22156 removes the need for the extra props here, so after that lands I should be able to tidy this up nicely and wrap it up.

merged 👍

deBhal commented:
Ok, I've rebased this on an up-to-date-master and added a few minor refinements and it's ready for review again.

The result seems pretty clean, but it was an interesting rebase, so keep a keen eye out for stupid.

ramonjd commented on client/components/domains/registrant-extra-info/style.scss:
```diff
@@ -31,6 +20,12 @@
 	}
+
+	.form-label .form-label__optional {
+		font-weight: 200;
```
To my eye at least, you could remove this font-weight declaration. The weight appears too undernourished compared with the page's overall typographical styles.

<img width="556" alt="screen shot 2018-02-28 at 4 24 01 pm" src="https://user-images.githubusercontent.com/6458278/36824033-7085ef46-1d54-11e8-9d0e-378b259e0e8b.png">

<img width="584" alt="screen shot 2018-02-28 at 4 24 09 pm" src="https://user-images.githubusercontent.com/6458278/36824036-733ea304-1d54-11e8-81b2-57781e6a4370.png">

ramonjd commented:
Noticed some other things:

The checkbox error message is bold due to the weight of 600 on the label element. Could be solved by adding `font-weight: normal` to `div.form-input-validation`:

<img width="366" alt="screen shot 2018-03-01 at 1 33 00 pm" src="https://user-images.githubusercontent.com/6458278/36824297-adcf4dec-1d55-11e8-8ba7-107705e75009.png">

Also, it's kind of cheeky, but I can get through the LLC check just by clicking quickly :) I'm just documenting it so you know. Not sure how important this is for `.ca`. 

![Uploading Mar-01-2018 13-42-14.gif…]()

deBhal commented:
> Not sure how important this is for .ca

It's important enough that I meant to fix it before merging this, but I see now that that was wrong and I'll merge this and open a new PR.

@ramonjd: I've addressed all feedback and fixed an error that snuck in through the rebased - want to give it a quick check before I deploy?

ramonjd commented:
LGTM 👍 

As done previously, I have retested adding `.ca` domain and a `.uk` for good measure :)
